### PR TITLE
Configurable Kubernetes version setup

### PR DESCRIPTION
**Pull request by JulienBalestra**
### What does this PR do?

This PR allows to choose a Kubernetes (hyperkube) version and have a working setup by default.
To achieve that in a reliable way, I had to move the kube-apiserver from static pod to systemd.

Kubernetes versions:
- [x] 1.11
- [x] 1.10
- [x] 1.9
- [x] 1.8
- [x] 1.7
- [x] 1.6
- [x] 1.5

### Motivation

Be able to quickly test any new Kubernetes version, and still be compatible with the old ones.

### Additional Notes

The systemd unit is now rendered in two phases.
The cleanup stage now stops systemd units if they're already started.

It's now possible to select all options but one with:
```
# ./pupernetes daemon run sandbox/ -c -binaries

Clean options are: {"etcd":true,"manifests":true,"kubelet":true,"secrets":true,"network":true,"systemd":true,"kubectl":true,"mounts":true,"iptables":true}
```

The draining phase is also revamped to depend on `/var/log/pods` to determine if the kubelet GC is passed.

### CODE REVIEW

**Comment by JulienBalestra** [comment]
Rebased on master after #32

File: README.md [hunk 1]
```diff
@@ -17,15 +25,16 @@ Run a Kubernetes setup in 45 seconds.
 
 The default setup is secured with:
 
-* x509 certificates provided by an embedded vault PKI
+* valid x509 certificates provided by an embedded vault PKI
+    * Able to use the Kubernetes CSR and the service account root-ca
```
Comment (mfpierre) [inline README.md]: caps or no caps you have to choose 😬

File: README.md [hunk 2]
```diff
@@ -1,6 +1,14 @@
 # pupernetes - p8s
 
-Run a Kubernetes setup in 45 seconds.
+Run a managed Kubernetes setup.
+
+This project purpose is to provide a simple Kubernetes setup to validate any software on top of it.
+
+You can use it to validate a software dependence on Kubernetes itself or just to run some classic app workflows with [argo](https://github.com/argoproj/argo)
+
+Our main use-case is the end to end testing pipeline of the [datadog-agent](https://github.com/DataDog/datadog-agent)
```
Comment (mfpierre) [inline README.md]: s/use-case/use case

File: README.md [hunk 3]
```diff
@@ -60,7 +69,7 @@ Setup a linux environment for running `pupernetes`. **This is only a suggested e
 
 #### Install VMWare Fusion
 
-`pupernetes` must be run on linux. To run a linux VM install [VMWare Fusion](https://www.vmware.com/products/fusion/fusion-evaluation.html) or your preferred virtualization software. You can use the VMWare Fusion Pro 30-day trial.
+`pupernetes` must be run on linux. To run a linux VM install [VMware Fusion](https://www.vmware.com/products/fusion/fusion-evaluation.html) or your preferred virtualization software. You can use the VMware Fusion Pro 30-day trial.
```
Comment (mfpierre) [inline README.md]: IMO we should just say "must be run on linux (or linux VM)"

File: pkg/run/probe.go [hunk 4]
```diff
@@ -28,3 +29,29 @@ func (r *Runtime) httpProbe(url string) error {
...
+		glog.V(3).Infof("%s", s)
+		switch u.SubState {
+		case "running":
+			continue
+		case "start":
+			continue
+		}
+		glog.Errorf("Unexpected state of: %s", s)
+		failed = append(failed, u.Name)
+
```
Comment (mfpierre) [inline pkg/run/probe.go]: remove extra line

File: pkg/setup/systemd.go [hunk 5]
```diff
@@ -175,140 +177,22 @@ func (e *Environment) createEnd2EndSection() []*unit2.UnitOption {
...
-		},
+	defer fd.Close()
+	unitOptions, err := unit2.Deserialize(fd)
+	if err != nil {
+		glog.Errorf("Unexpected error during parsing s: %v", unitNameNoPrefix, err)
+		return err
 	}
-	sdKubelet = append(sdKubelet, e.systemdEnd2EndSection...)
-	err := e.writeSystemdUnit(sdKubelet, fmt.Sprintf("%setcd.service", config.ViperConfig.GetString("systemd-unit-prefix")))
+	//unitOptions = append(unitOptions, e.systemdEnd2EndSection...) TODO see how to insert that
```
Comment (mfpierre) [inline pkg/setup/systemd.go]: you can leave the comment but remove the code

**Pull request merged**